katip-elasticsearch/src/Katip/Scribes/ElasticSearch/Internal.hs

```diff
@@ -518,3 +510,156 @@ instance ESVersion ESV5 where
+    mapM_ waitCatch workers
+    putMVar endSig ()
+
+  let finalizer = putMVar endSig () >> takeMVar endSig
+  return (Scribe (logger inChan) finalizer)
+  where
+    logger :: forall a
+            . LogItem a
+           => UNB.InChan (IndexName v, Value)
+           -> Item a
+           -> IO ()
+    logger inChan i = when (_itemSeverity i >= sev) $
+      UNB.writeChan inChan (chooseIxn prx ix essIndexSharding i, itemJson' i)
+
+    prx :: Typeable.Proxy v
+    prx = Typeable.Proxy
+
+    tplName = toTemplateName prx ixn
+
+    shardingEnabled = case essIndexSharding of
+      NoIndexSharding -> False
+      _               -> True
+
+    tpl = toIndexTemplate prx (toTemplatePattern prx (ixn <> "-*")) (Just essIndexSettings) [toJSON base]
+
+    base = baseMapping prx mapping
+
+    ixn = fromIndexName prx ix
+
+    itemJson' :: LogItem a => Item a -> Value
+    itemJson' i
+      | essAnnotateTypes = itemJson verb (TypeAnnotated <$> i)
+      | otherwise        = itemJson verb i
+
+
+data DiscrimLen =
+    SendIt
+  | StackIt
+  deriving Show
+
+startBulkWorker
+    :: forall v . (ESVersion v, V5.MonadBH (BH v IO))
+    => EsScribeCfg v
+    -> BHEnv v
+    -> MappingName v
+    -> MVar ()
+    -> UNB.OutChan (IndexName v, Value)
+    -> IO ()
+startBulkWorker _ env mapping dieSignal outChan = do
+  -- We need to randomize upload delay
+  -- so that workers do not stampede the nodes
+  delayTime' <- randomRIO (100, 2000)
+  go (delayTime' * milliseconds) [] 0
+  where
+    milliseconds :: Int
+    milliseconds = 1000
+
+    go :: Int -> [V5.BulkOperation] -> Int -> IO ()
+    go delayTime xs len = do
+      result <- race (UNB.readChan outChan) (threadDelay delayTime)
+      case result of
+        (Left val) ->
+          case discrimLen len of
+            SendIt -> do
+              newPair <- serializePair val
+              uploadData (newPair : xs)
+              dieOrContinue [] 0
+            StackIt -> do
+              newPair <- serializePair val
+              dieOrContinue (newPair : xs) (len+1)
+        (Right ()) -> do
+          uploadData xs
+          dieOrContinue [] 0
+      where dieOrContinue newList newLen = do
+              dieNow <- tryReadMVar dieSignal
+              case dieNow of
+                Nothing -> go delayTime newList newLen
+                (Just ()) -> return ()
+
+    uploadData :: [V5.BulkOperation] -> IO ()
+    uploadData [] = return ()
+    uploadData xs = do
+      reply <- runBH prx env $ V5.bulk $ V.fromList xs
+      case statusCode $ responseStatus reply of
+        200 -> return ()
+        s -> print s
+
+    serializePair :: (IndexName v, Value) -> IO V5.BulkOperation
+    serializePair (indexName, val) = do
+      docId <- mkDocId'
+      return $
+        V5.BulkIndex
+        (V5.IndexName (fromIndexName prx indexName))
+        (V5.MappingName (fromMappingName prx mapping))
+        (V5.DocId docId) val
+
+    discrimLen l
+      | l >= 500 = SendIt
```

Comment:
I wonder if we should make this configurable.

Reply:
The figure was the product of throughput testing. I don't know why someone would want to change it, but yeah probably should be configurable.